### PR TITLE
A consistent interface for plugin extended functionality

### LINKED ISSUE #83244: [framework] a consistent interface for plugin extended functionality
**What would you like to be added**:
A consistent way for adding extended functionality to plugins.

**Why is this needed**:
Currently we have two types of plugins that support adding some sort of optional functionality: PreFilter allow adding "AddPod/RemovePod" implementations through an "Updater" interface, while Score includes a NormalizeScore, which must be implemented, but plugins that don't require it can return nil.

The two cases above support adding "optional" functionality in different ways, what I suggest we do is have a common pattern, with common naming conventions to support that:

Any plugin that needs an extended functionality should define an ```_PluginName_Extensions``` interface, and have a method named ```Extensions()``` that returns the interface or nil if the plugin does not implement those extensions.

If we do that for PreFilter and Score plugins, it would look like this:

```
type PreFilterExtensions interface {
	AddPod(pc *PluginContext, podToSchedule *v1.Pod, podToAdd *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status
	RemovePod(pc *PluginContext, podToSchedule *v1.Pod, podToRemove *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status
}

type PreFilterPlugin interface {
	Plugin
	PreFilter(pc *PluginContext, p *v1.Pod) *Status
	Extensions() PreFilterExtensions
}


type ScoreExtensions interface {
    NormalizeScore(pc *PluginContext, p *v1.Pod, scores NodeScoreList) *Status
}

type ScorePlugin interface {
	Plugin
        Score(pc *PluginContext, p *v1.Pod, nodeName string) (int, *Status)
	Extensions() PreFilterExtensions
}
```

One concern here is that if the plugin wants to implement a single method of the extensions interface, they it will have to implement all the extensions methods, but i think that is fine since we don't expect to have many methods anyways.

### PR DESCRIPTION
**Pull request by notpad**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

A consistent way for adding extended functionality to plugins.

**Which issue(s) this PR fixes**:

Fixes #83244

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
None
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/v1alpha1/framework.go [hunk 1]
```diff
@@ -417,11 +417,13 @@ func (f *framework) RunScorePlugins(pc *PluginContext, pod *v1.Pod, nodes []*v1.
 	workqueue.ParallelizeUntil(ctx, 16, len(f.scorePlugins), func(index int) {
 		pl := f.scorePlugins[index]
 		nodeScoreList := pluginToNodeScores[pl.Name()]
-		status := pl.NormalizeScore(pc, pod, nodeScoreList)
-		if !status.IsSuccess() {
-			err := fmt.Errorf("normalize score plugin %q failed with error %v", pl.Name(), status.Message())
-			errCh.SendErrorWithCancel(err, cancel)
-			return
+		if extensions := pl.Extensions(); extensions != nil {
```
Comment (damemi) [inline pkg/scheduler/framework/v1alpha1/framework.go]: Maybe you can combine this check with the one above in https://github.com/kubernetes/kubernetes/pull/83318/files#diff-76b7e314f9618c8bc2ca6442ddd35dc3R326? Might make it clearer that these are the same interface, especially since there it's just `extension` while this is`extensions`

File: pkg/scheduler/framework/v1alpha1/framework.go [hunk 2]
```diff
@@ -317,14 +317,14 @@ func (f *framework) RunPreFilterUpdaterAddPod(pc *PluginContext, podToSchedule *
...
+// RunPreFilterExtensionRemovePod calls the RemovePod interface for the set of configured
 // PreFilter plugins. It returns directly if any of the plugins return any
 // status other than Success.
-func (f *framework) RunPreFilterUpdaterRemovePod(pc *PluginContext, podToSchedule *v1.Pod,
+func (f *framework) RunPreFilterExtensionRemovePod(pc *PluginContext, podToSchedule *v1.Pod,
 	podToRemove *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status {
 	for _, pl := range f.preFilterPlugins {
-		if updater := pl.Updater(); updater != nil {
-			status := updater.RemovePod(pc, podToSchedule, podToRemove, nodeInfo)
+		if extension := pl.Extensions(); extension != nil {
```
Comment (damemi) [inline pkg/scheduler/framework/v1alpha1/framework.go]: `extensions` plural instead?

File: pkg/scheduler/framework/v1alpha1/interface.go [hunk 3]
```diff
@@ -230,6 +230,13 @@ type PostFilterPlugin interface {
 	PostFilter(pc *PluginContext, pod *v1.Pod, nodes []*v1.Node, filteredNodesStatuses NodeToStatusMap) *Status
 }
 
+type ScoreExtensions interface {
```
Comment (damemi) [inline pkg/scheduler/framework/v1alpha1/interface.go]: Need a comment here to fix golint failure

File: pkg/scheduler/framework/v1alpha1/interface.go [hunk 4]
```diff
@@ -239,13 +246,9 @@ type ScorePlugin interface {
...
-	// NormalizeScore is called for all node scores produced by the same plugin's "Score"
-	// method. A successful run of NormalizeScore will update the scores list and return
-	// a success status.
-	//
-	// NOTE: This function is optional, and you could implement it as a no-op by simply
-	// returning nil.
-	NormalizeScore(pc *PluginContext, p *v1.Pod, scores NodeScoreList) *Status
+	// Extensions returns an extension if the plugin implements one, or nil if it
+	// does not.
+	Extensions() ScoreExtensions
```
Comment (draveness) [inline pkg/scheduler/framework/v1alpha1/interface.go]: I think `Extension` is more appropriate here to me

File: pkg/scheduler/framework/v1alpha1/framework.go (hunk 2 above)
Comment (damemi) [reply pkg/scheduler/framework/v1alpha1/framework.go]: really just looking for consistency

File: pkg/scheduler/framework/v1alpha1/interface.go (hunk 4 above)
Comment (ahg-g) [reply pkg/scheduler/framework/v1alpha1/interface.go]: I would keep it as ```Extensions```, the interface may include one or more extension (AddPod and RemovePod for example).

File: pkg/scheduler/framework/v1alpha1/framework.go [hunk 5]
```diff
@@ -297,14 +297,14 @@ func (f *framework) RunPreFilterPlugins(
...
+// RunPreFilterExtensionAddPod calls the AddPod interface for the set of configured
 // PreFilter plugins. It returns directly if any of the plugins return any
 // status other than Success.
-func (f *framework) RunPreFilterUpdaterAddPod(pc *PluginContext, podToSchedule *v1.Pod,
+func (f *framework) RunPreFilterExtensionAddPod(pc *PluginContext, podToSchedule *v1.Pod,
 	podToAdd *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status {
 	for _, pl := range f.preFilterPlugins {
-		if updater := pl.Updater(); updater != nil {
-			status := updater.AddPod(pc, podToSchedule, podToAdd, nodeInfo)
+		if extensions := pl.Extensions(); extensions != nil {
```
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/framework.go]: may be you don't even need to assign to a different variable, you can do something like this in the for loop to reduce indentation:

```
if pl.Extensions() == nil {
 continue;
}

if status := extension.RemovePod(pc, podToSchedule, podToRemove, nodeInfo); !status.IsSuccess() {
  msg := fmt.Sprintf("error while running RemovePod for plugin %q while scheduling pod %q: %v",
                     pl.Name(), podToSchedule.Name, status.Message())
  klog.Error(msg)
  return NewStatus(Error, msg)
}
```

File: pkg/scheduler/framework/v1alpha1/interface.go (hunk 4 above)
Comment (damemi) [reply pkg/scheduler/framework/v1alpha1/interface.go]: >  the interface may include one or more extension (AddPod and RemovePod for example).

that's how I interpreted it too

**Review by ahg-g — COMMENTED** [review]
Thanks, looking forward to merging this when done.

File: pkg/scheduler/framework/v1alpha1/framework.go (hunk 2 above)
Comment (notpad) [reply pkg/scheduler/framework/v1alpha1/framework.go]: Sorry for my careless,  I have changed extension to extensions.

File: pkg/scheduler/framework/v1alpha1/interface.go (hunk 3 above)
Comment (notpad) [reply pkg/scheduler/framework/v1alpha1/interface.go]: Fixed

File: pkg/scheduler/framework/v1alpha1/framework.go (hunk 5 above)
Comment (notpad) [reply pkg/scheduler/framework/v1alpha1/framework.go]: Thanks, very good suggestion~

File: pkg/scheduler/framework/v1alpha1/framework_test.go [hunk 6]
```diff
@@ -29,12 +29,12 @@ import (
...
-	scorePlugin1                   = "score-plugin-1"
-	pluginNotImplementingScore     = "plugin-not-implementing-score"
-	preFilterPluginName            = "prefilter-plugin"
-	preFilterWithUpdaterPluginName = "prefilter-with-updater-plugin"
+	scoreWithNormalizePlugin1        = "score-with-normalize-plugin-1"
+	scoreWithNormalizePlugin2        = "score-with-normalize-plugin-2"
+	scorePlugin1                     = "score-plugin-1"
+	pluginNotImplementingScore       = "plugin-not-implementing-score"
+	preFilterPluginName              = "prefilter-plugin"
+	preFilterWithExtensionPluginName = "prefilter-with-extension-plugin"
```
[the remark this answers is not in the document]
Comment (notpad) [reply pkg/scheduler/framework/v1alpha1/framework_test.go]: Replaced.

**Comment by ahg-g** [comment]
Thanks, please squash the commits.

File: pkg/scheduler/framework/v1alpha1/interface.go [hunk 7]
```diff
@@ -230,6 +230,15 @@ type PostFilterPlugin interface {
 	PostFilter(pc *PluginContext, pod *v1.Pod, nodes []*v1.Node, filteredNodesStatuses NodeToStatusMap) *Status
 }
 
+// ScoreExtensions is an interface that is included in plugins that allow specifying
+// callbacks to make additional change to the score list.
```
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/interface.go]: I would just say: ```ScoreExtensions is an interface for Score extended functionality.```

File: pkg/scheduler/framework/v1alpha1/interface.go [hunk 8]
```diff
@@ -239,13 +248,9 @@ type ScorePlugin interface {
...
 
-	// NormalizeScore is called for all node scores produced by the same plugin's "Score"
-	// method. A successful run of NormalizeScore will update the scores list and return
-	// a success status.
-	//
-	// NOTE: This function is optional, and you could implement it as a no-op by simply
-	// returning nil.
-	NormalizeScore(pc *PluginContext, p *v1.Pod, scores NodeScoreList) *Status
+	// Extensions returns an extension if the plugin implements one, or nil if it
+	// does not.
```
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/interface.go]: I would phrase it: "returns a ScoreExtensions interface if it implements one, or nil if does not"

**Comment by notpad** [comment]
Open a new pull request to squash the commits

File: pkg/scheduler/framework/v1alpha1/interface.go [hunk 9]
```diff
@@ -184,13 +184,13 @@ type PreFilterPlugin interface {
...
-	// does not. A Pre-filter plugin can provide an updater to incrementally
-	// modify its pre-processed info. The framework guarantees that the updater
+	// Extensions returns an extension if the plugin implements one, or nil if it
+	// does not. A Pre-filter plugin can provide an extension to incrementally
+	// modify its pre-processed info. The framework guarantees that the extension
 	// AddPod/RemovePod functions will only be called after PreFilter,
 	// possibly on a cloned PluginContext, and may call those functions more than
 	// once before calling Filter again on a specific node.
-	Updater() Updater
+	Extensions() PreFilterExtensions
```
Comment (ahg-g) [inline pkg/scheduler/framework/v1alpha1/interface.go]: I would rephrase the comment like this:

Extensions returns an extensions interface if the plugin implements one, or nil if it does not. A Pre-filter plugin can provide extensions to incrementally modify its pre-processed info. The framework guarantees that the extensions AddPod/RemovePod will only be called after PreFilter,	possibly on a cloned PluginContext, and may call those functions more than once before calling Filter again on a specific node.

**Review by ahg-g — COMMENTED** [review]
few suggestions regarding how we document Extensions.

**Comment by ahg-g** [comment]
> Open a new pull request to squash the commits

We need to keep the review history, so you don't need a new pull request to squash the commits, you can just do the following:

git reset HEAD~${numberOfCommits} 
git commit -m "new commit message"
git push -f

where ${numberOfCommits} is the number of commits you applied, in this case it seems to be five.

/open

**Comment by notpad** [comment]
The commits include a merge, so this seems to not work well. I've force pushed the code, but the pull request can not be opened now.

New pull request is opened here: https://github.com/kubernetes/kubernetes/pull/83365

**Comment by ahg-g** [comment]
> The commits include a merge, so this seems to not work well. I've force pushed the code, but the pull request can not be opened now.
>

hmm, it seems something wrong happened when you rebased, ok lets continue on the other PR, but next time please make sure that things work out on the same PR.
